src/dogroups.c

```diff
@@ -177,45 +180,45 @@ SEXP dogroups(SEXP dt, SEXP dtcols, SEXP groups, SEXP grpcols, SEXP jiscols, SEX
     } else {
       if (LOGICAL(verbose)[0]) tstart = clock();
       SETLENGTH(I, grpn);
+      int *iI = INTEGER(I);
       if (LENGTH(order)==0) {
-        if (grpn) rownum = INTEGER(starts)[i]-1; else rownum = -1;  // not ternary to pass strict-barrier
-        for (j=0; j<grpn; j++) INTEGER(I)[j] = rownum+j+1;
+        if (grpn) rownum = istarts[i]-1; else rownum = -1;  // not ternary to pass strict-barrier
+        for (j=0; j<grpn; j++) iI[j] = rownum+j+1;
         if (rownum>=0) {
           for (j=0; j<length(SDall); j++) {
             size_t size = SIZEOF(VECTOR_ELT(SDall,j));
-            memcpy((char *)DATAPTR(VECTOR_ELT(SDall,j)),  // direct memcpy best here, for usually large size groups. by= each row is slow and not recommended anyway, so we don't mind there's no switch here for grpn==1
+            memcpy((char *)DATAPTR(VECTOR_ELT(SDall,j)),  // direct memcpy best here, for usually large size groups. by= each row is slow and not recommended anyway, so we don't mind there's no switch here for grpn==1 // TODO remove DATAPTR
                    (char *)DATAPTR(VECTOR_ELT(dt,INTEGER(dtcols)[j]-1))+rownum*size,
                    grpn*size);
             // SD is our own alloc'd memory, and the source (DT) is protected throughout, so no need for SET_* overhead
           }
           for (j=0; j<length(xSD); j++) {
             size_t size = SIZEOF(VECTOR_ELT(xSD,j));
-            memcpy((char *)DATAPTR(VECTOR_ELT(xSD,j)),  // ok use of memcpy. Loop'd through columns not rows
+            memcpy((char *)DATAPTR(VECTOR_ELT(xSD,j)),  // ok use of memcpy. Loop'd through columns not rows // TODO remove DATAPTR
```

Comment:
in cases like this we have to `switch` and case by case? or is there any better way?

Reply:
maybe call `memrecycle`